Test dynamo wrapper

mohit opened the pull request:
A prototype for get/set results to DyanamoDB. Supports some functionality of `gearcmd`.

No tests yet since not currently sure if this is the model we will be taking.

Commit:
default template files

Commit:
prototype command wrapper with dynamo input/output

Commit:
add Dockerfile and push image to dockerhub

Commit:
docker image on branch

Commit:
use full patch for entrypoint

Commit:
add logging and actually use main/main

Commit:
local mode for testing. arg parsing for -cmd

Commit:
hardcode session to us-east-1

Commit:
install ca-certificates in Dockerfile

Commit:
add: query results from depedencies

rgarcia commented on README.md:
````diff
@@ -1,2 +1,15 @@
+## Usage
+
+```
+gearcmd -name <job id> -input <key-id> -cmd <command>
````
s/gearcmd/batchcli/

rgarcia commented on runner/batch.go:
```diff
@@ -0,0 +1,49 @@
+	Queue              string
+	ComputeEnvironment string
+	DependencyIds      []string
+}
```
without a good understanding of AWS Batch, it's hard to get oriented on what this struct represents. Comments on this struct + field values would make it easier!

rgarcia commented on runner/batch.go:
```diff
@@ -0,0 +1,49 @@
+}
+
+// NewBatchJobFromEnv returns a new BatchJob by reading
+// values from the envrionment
```
s/envrionment/environment/

rgarcia commented on runner/batch.go:
```diff
@@ -0,0 +1,49 @@
+			errors.New("AWS Batch environment variables not found")
+	}
+
+	depsStr := os.Getenv("_BATCH_DEPEDENCIES")
```
comment about why this environment variable is different from the others

rgarcia commented on runner/runner.go:
```diff
@@ -0,0 +1,71 @@
+	cmd.Env = os.Environ()
+
+	// Write the stdout and stderr of the process to both this process' stdout and stderr
+	// and also write it to a byte buffer so that we can return it with the Gearman job
```
mentions Gearman

rgarcia commented on runner/runner.go:
```diff
@@ -0,0 +1,71 @@
+	// example:
+	// 		batchcli -cmd echo hello there
+	// 		results = [{"json":"true"}, {}]
+	//      exec(echo, ["hello", "there", '{"json":"true"}', '{}'])
```
interesting approach to getting the results of dependent jobs... Is there a way for the user to know which positional argument belongs to which dependency, or will they have to have an assumption around dependency ordering?

mohit replied:
Not sure yet. I've been wondering if I should pass in the dependent job-name (which I don't have access to without additional env vars) or just the dependency job-ids.

One other direction I had considered is to write the job-name to dyanmo and then pull out results as a map:

```bash
batchcli -cmd echo hello there
results = {"job-one-name": {"json":"true"}, "job-two-name": {}}

exec(echo, ["hello", "there", {"job-one-name": {"json":"true"}, "job-two-name": {}}]
```
This is simpler, but now it means that there is a dependency between job-names in the workflow manager and the Worker code.  

Another option would be to just pass the results as a list `[]` and the underlying Worker should figure out which object is from what dependency.

rgarcia commented on runner/store.go:
```diff
@@ -0,0 +1,101 @@
+		return DynamoStore{}, err
+	}
+
+	client := dynamodb.New(sess)
```
IMO constructor here should take in [`dynamodbiface.DynamoDBAPI`](https://godoc.org/github.com/aws/aws-sdk-go/service/dynamodb/dynamodbiface#DynamoDBAPI) and the initialization of the session and client object should happen in callers of `NewDynamoStore()`.

rgarcia commented on runner/store.go:
```diff
@@ -0,0 +1,101 @@
+				S: aws.String(key),
+			},
+			"Result": {
+				S: aws.String(result),
```
probably want to put a check here for limits around key/val and total item size in dynamo, since this is the raw stdout or stderr

rgarcia reviewed (commented):
did an initial pass, looks nice. Would be nice to comment on the Batch-specific stuff in here. Besides that the storing results in dynamo code was pretty easy to follow.

Will this eventually interface with the Batch API in some way, or is its sole purpose capturing / storing results in a standard way?

mohit commented:
Thanks. Exactly the kinds of comments that I was looking for. This should ideally not interact with the AWS Batch API. It's only a wrapper to allow for I/O capture and for passing results between jobs.

Commit:
Add comments and fix README; use dynamodbiface

mohit commented:
@rgarcia thanks again for the feedback. Added comments and fixed references I had copied over from gearcmd. Still not sure about how to best pass outputs of dependencies over command line but I'm hoping to figure that out over the rest of this week. https://github.com/Clever/batchcli/pull/1/commits/76c2adc9f45b8190afd89776322f80d3a8690044

Commit:
typofix
